File: frontend/src/lib/fileedit/diff.ts
// Shared diff engine (TS port of internal/fileedit/diff.go).
//
// ONE algorithm, TWO implementations: the Go engine computes the ledger's
// ground-truth unified diffs server-side; this TS port computes the SAME
// bytes for the SAME snapshot pair so the GUI/TUI renderers provably display
// exactly what the ledger stores. Parity is enforced by the shared test
// vectors in internal/testfixtures/fileedit/*.json, which BOTH engines must
// reproduce byte-for-byte (frontend/src/lib/fileedit/testvectors.test.ts vs
// internal/fileedit/diff_test.go).
//
// Conventions (identical to the Go engine):
//   - unified diff, 3 context lines, git-style a/ b/ headers (/dev/null for
//     pure creates/deletes), LF endings
//   - "\ No newline at end of file" markers
//   - hunk grouping: two changes share a hunk while the unchanged gap is
//     <= 2*ctx lines
//   - binary = NUL byte in the first 8 KiB (git's heuristic) → empty diff
//   - byte-identical before/after → empty diff (no ledger entry)
//   - diff text capped at 256 KiB (truncated flag)
//   - UTF-8 passes through untouched: strictly line-level diffing

export const MAX_DIFF_BYTES = 256 << 10; // 256 KiB
const BINARY_PROBE_LEN = 8192;
const DIFF_CONTEXT = 3;

export type DiffKind = "create" | "modify" | "delete" | "";

export interface DiffResult {
  unifiedDiff: string;
  kind: DiffKind;
  binary: boolean;
  truncated: boolean;
}

interface DLine {
  text: string;
  hasEOL: boolean;
}

// op is one edit-script operation: ' ' keep, '-' remove, '+' add.
interface DOp {
  kind: "|" | "-" | "+" | " ";
  aIdx: number;
  bIdx: number;
}

const enc = new TextEncoder();

function isBinary(content: string | null): boolean {
  if (content === null || content.length === 0) return false;
  const bytes = enc.encode(
    content.length > BINARY_PROBE_LEN * 4
      ? content // encode() is linear; the probe cap matters for byte windows
      : content,
  );
  const end = Math.min(bytes.length, BINARY_PROBE_LEN);
  for (let i = 0; i < end; i++) {
    if (bytes[i] === 0) return true;
  }
  return false;
}

// splitLines splits content into diffable lines — the exact convention of
// the Go engine: a trailing "\n" does NOT produce an extra empty line, and
// the final fragment without a newline keeps hasEOL=false.
function splitLines(c: string): DLine[] {
  if (c.length === 0) return [];
  const endsNL = c.endsWith("\n");
  const parts = c.split("\n");
  if (endsNL) parts.pop();
  const out: DLine[] = new Array(parts.length);
  for (let i = 0; i < parts.length; i++) {
    out[i] = { text: parts[i], hasEOL: i < parts.length - 1 || endsNL };
  }
  return out;
}

function kindFor(beforeExisted: boolean, afterExisted: boolean): DiffKind {
  if (!beforeExisted && afterExisted) return "create";
  if (beforeExisted && !afterExisted) return "delete";
  return "modify";
}

// diffOps computes a minimal edit script via Myers' O((N+M)D) greedy
// algorithm with common prefix/suffix trimming — the Go engine's exact
// approach, so both sides emit the same (minimal) edit script.
function diffOps(a: DLine[], b: DLine[]): DOp[] {
  const n = a.length;
  const m = b.length;
  let p = 0;
  while (p < n && p < m && a[p].text === b[p].text) p++;
  let s = 0;
  while (s < n - p && s < m - p && a[n - 1 - s].text === b[m - 1 - s].text) s++;
  const ops: DOp[] = [];
  for (let i = 0; i < p; i++) ops.push({ kind: " ", aIdx: i, bIdx: i });
  ops.push(...opsMyers(a.slice(p, n - s), b.slice(p, m - s), p));
  for (let i = 0; i < s; i++) {
    ops.push({ kind: " ", aIdx: n - s + i, bIdx: m - s + i });
  }
  return ops;
}

// opsMyers runs Myers' greedy algorithm on the middle section with the same
// effort guard as the Go engine: when the edit distance explodes (a
// near-total rewrite), fall back to the naive whole-middle replace — the
// same script a minimal diff would render anyway.
function opsMyers(a: DLine[], b: DLine[], prefix: number): DOp[] {
  const n = a.length;
  const m = b.length;
  if (n === 0 && m === 0) return [];
  if (n === 0) {
    const ops: DOp[] = [];
    for (let i = 0; i < m; i++) ops.push({ kind: "+", aIdx: 0, bIdx: prefix + i });
    return ops;
  }
  if (m === 0) {
    const ops: DOp[] = [];
    for (let i = 0; i < n; i++) ops.push({ kind: "-", aIdx: prefix + i, bIdx: 0 });
    return ops;
  }

  const max = n + m;
  const offset = max;
  if (max > 200_000) return opsNaive(a, b, prefix);
  const v = new Int32Array(2 * max + 1);
  const trace: Int32Array[] = [];
  const dCap = Math.max(16, Math.floor((32 << 20) / (16 * max + 8)));

  let found = false;
  for (let d = 0; d <= max; d++) {
    if (d > dCap) return opsNaive(a, b, prefix);
    trace.push(v.slice());
    for (let k = -d; k <= d; k += 2) {
      let x: number;
      if (k === -d || (k !== d && v[offset + k - 1] < v[offset + k + 1])) {
        x = v[offset + k + 1];
      } else {
        x = v[offset + k - 1] + 1;
      }
      let y = x - k;
      while (x < n && y < m && a[x].text === b[y].text) {
        x++;
        y++;
      }
      v[offset + k] = x;
      if (x >= n && y >= m) {
        found = true;
        break;
      }
    }
    if (found) break;
  }

  // Backtrack the trace to reconstruct the (reversed) edit path.
  const rev: DOp[] = [];
  let x = n;
  let y = m;
  for (let d = trace.length - 1; d >= 1; d--) {
    const vv = trace[d];
    const k = x - y;
    let prevK: number;
    if (k === -d || (k !== d && vv[offset + k - 1] < vv[offset + k + 1])) {
      prevK = k + 1;
    } else {
      prevK = k - 1;
    }
    const prevX = vv[offset + prevK];
    const prevY = prevX - prevK;
    while (x > prevX && y > prevY) {
      x--;
      y--;
      rev.push({ kind: " ", aIdx: x + prefix, bIdx: y + prefix });
    }
    if (x === prevX && y === prevY) continue;
    if (x > prevX) {
      x--;
      rev.push({ kind: "-", aIdx: x + prefix, bIdx: 0 });
    } else {
      y--;
      rev.push({ kind: "+", aIdx: 0, bIdx: y + prefix });
    }
  }
  const ops: DOp[] = new Array(rev.length);
  for (let i = 0; i < rev.length; i++) ops[i] = rev[rev.length - 1 - i];
  return ops;
}

// opsNaive is the fallback edit script: remove the whole middle-A block,
// add the whole middle-B block. Linear time, always correct.
function opsNaive(a: DLine[], b: DLine[], prefix: number): DOp[] {
  const ops: DOp[] = [];
  for (let i = 0; i < a.length; i++) ops.push({ kind: "-", aIdx: prefix + i, bIdx: 0 });
  for (let i = 0; i < b.length; i++) ops.push({ kind: "+", aIdx: 0, bIdx: prefix + i });
  return ops;
}

// hunkRange renders one side of a hunk header: bare start when count==1
// (git's convention), "start,count" otherwise.
function hunkRange(start: number, count: number): string {
  return count === 1 ? `${start}` : `${start},${count}`;
}

function writeDline(l: DLine, sign: string): string {
  return l.hasEOL ? sign + l.text + "\n" : `${sign}${l.text}\n\\ No newline at end of file\n`;
}

// writeHunks renders the ops into standard hunks (3 context lines, paired
// line numbers) — the Go engine's grouping rule verbatim.
function writeHunks(a: DLine[], b: DLine[]): string {
  const ops = diffOps(a, b);
  const ctx = DIFF_CONTEXT;
  let out = "";

  type Hunk = { from: number; to: number };
  const hunks: Hunk[] = [];
  let i = 0;
  while (i < ops.length) {
    if (ops[i].kind === " ") {
      i++;
      continue;
    }
    const from = Math.max(0, i - ctx);
    let to = i + 1;
    while (to < ops.length) {
      if (ops[to].kind !== " ") {
        to++;
        continue;
      }
      let j = to;
      while (j < ops.length && ops[j].kind === " ") j++;
      if (j < ops.length && j - to <= 2 * ctx) {
        to = j + 1;
        continue;
      }
      break;
    }
    const end = Math.min(to + ctx, ops.length);
    hunks.push({ from, to: end });
    i = end;
  }

  for (const h of hunks) {
    let countA = 0;
    let countB = 0;
    let startA = 0;
    let startB = 0;
    let first = true;
    for (let o = h.from; o < h.to; o++) {
      const op = ops[o];
      if (first) {
        startA = op.aIdx + 1;
        startB = op.bIdx + 1;
        first = false;
      }
      if (op.kind === " ") {
        countA++;
        countB++;
      } else if (op.kind === "-") {
        countA++;
      } else {
        countB++;
      }
    }
    out += `@@ -${hunkRange(startA, countA)} +${hunkRange(startB, countB)} @@\n`;
    for (let o = h.from; o < h.to; o++) {
      const op = ops[o];
      if (op.kind === " ") out += writeDline(a[op.aIdx], " ");
      else if (op.kind === "-") out += writeDline(a[op.aIdx], "-");
      else out += writeDline(b[op.bIdx], "+");
    }
  }
  return out;
}

// writeWholeHunk renders the whole-file hunk for pure creates and deletes.
function writeWholeHunk(lines: DLine[], sign: "+" | "-"): string {
  let out =
    sign === "+"
      ? `@@ -0,0 +1,${lines.length} @@\n`
      : `@@ -1,${lines.length} +0,0 @@\n`;
  for (const l of lines) out += writeDline(l, sign);
  return out;
}

// computeUnifiedDiff is the TS twin of fileedit.ComputeUnifiedDiff: a
// ground-truth unified diff between two content snapshots. before==null
// means the file did not exist (create); after==null means deletion.
// Byte-identical pairs yield "" (no ledger entry).
export function computeUnifiedDiff(
  before: string | null,
  after: string | null,
  path: string,
): DiffResult {
  const beforeExisted = before !== null;
  const afterExisted = after !== null;
  if (before === null && after === null) {
    return { unifiedDiff: "", kind: "", binary: false, truncated: false };
  }
  if (beforeExisted && afterExisted && before === after) {
    return { unifiedDiff: "", kind: "", binary: false, truncated: false };
  }

  const kind = kindFor(beforeExisted, afterExisted);
  if (isBinary(before) || isBinary(after)) {
    return { unifiedDiff: "", kind, binary: true, truncated: false };
  }

  let header = `--- a/${path}\n+++ b/${path}\n`;
  if (!beforeExisted) header = `--- /dev/null\n+++ b/${path}\n`;
  else if (!afterExisted) header = `--- a/${path}\n+++ /dev/null\n`;

  let body = "";
  if (!beforeExisted) body = writeWholeHunk(splitLines(after as string), "+");
  else if (!afterExisted) body = writeWholeHunk(splitLines(before as string), "-");
  else body = writeHunks(splitLines(before as string), splitLines(after as string));

  let out = header + body;
  let truncated = false;
  if (out.length > MAX_DIFF_BYTES) {
    // The Go engine trims raw bytes; cut at a UTF-8-safe boundary.
    out = [...out].reduce((acc, ch) => (acc.length + ch.length <= MAX_DIFF_BYTES ? acc + ch : acc), "");
    truncated = true;
  }
  return { unifiedDiff: out, kind, binary: false, truncated };
}

File: internal/fileedit/diff.go
// Package fileedit implements the diff pipeline: server-computed, ground-truth
// unified diffs for every file a session edits. Diffs are computed from real
// file state (snapshot pairs / in-memory before+after), never parsed from tool
// output prose; the ledger rows persist to file_edit_ledger and reconcile
// against the run worktree's git state on completion.
package fileedit

import (
	"bytes"
	"fmt"
)

// Caps and detection thresholds (see architecture-notes/diff-pipeline-file-edit-ledger-server-computed-diffs.md §5).
const (
	// maxDiffBytes caps a single ledger entry's unified diff. Beyond it the
	// diff is cut (truncated=true) so one giant rewrite cannot blow up the
	// ledger, the event stream, or a renderer.
	maxDiffBytes = 256 << 10 // 256 KiB
	// binaryProbeLen is how many leading bytes are probed for a NUL to
	// classify content as binary — git's heuristic.
	binaryProbeLen = 8192
	// diffContext is the number of unchanged context lines shown around each
	// change in a hunk (git's -U3 default).
	diffContext = 3
)

// DiffKind enumerates the ledger entry kinds.
const (
	KindCreate = "create"
	KindModify = "modify"
	KindDelete = "delete"
)

// DiffResult is the outcome of ComputeUnifiedDiff for one file.
type DiffResult struct {
	// UnifiedDiff is the unified diff text ("--- a/<path>" …), 3 context
	// lines, LF line endings, empty for binary files.
	UnifiedDiff string
	// Kind is create | modify | delete.
	Kind string
	// Binary reports NUL-in-first-8KiB on either side (empty diff).
	Binary bool
	// Truncated reports the diff text was cut at maxDiffBytes.
	Truncated bool
}

// ComputeUnifiedDiff computes a ground-truth unified diff between two file
// content snapshots. before==nil means the file did not exist (create);
// after==nil means it was deleted. Byte-identical pairs yield "" (no entry):
// callers skip ledger rows for no-op edits. UTF-8 passes through untouched —
// diffing is strictly line-level, never byte- or rune-rewriting.
func ComputeUnifiedDiff(before, after []byte, path string) DiffResult {
	beforeExisted := before != nil
	afterExisted := after != nil
	if before == nil && after == nil {
		return DiffResult{}
	}
	if beforeExisted && afterExisted && bytes.Equal(before, after) {
		return DiffResult{} // no-op edit
	}

	kind := kindFor(beforeExisted, afterExisted)
	if isBinary(before) || isBinary(after) {
		return DiffResult{Kind: kind, Binary: true}
	}

	// Header: git-style a/ b/ prefixes; /dev/null for pure create/delete.
	header := fmt.Sprintf("--- a/%s\n+++ b/%s\n", path, path)
	if !beforeExisted {
		header = fmt.Sprintf("--- /dev/null\n+++ b/%s\n", path)
	} else if !afterExisted {
		header = fmt.Sprintf("--- a/%s\n+++ /dev/null\n", path)
	}

	var b bytes.Buffer
	b.WriteString(header)
	switch {
	case !beforeExisted:
		writeWholeHunk(&b, '+', splitLines(after)) // "@@ -0,0 +1,N @@"
	case !afterExisted:
		writeWholeHunk(&b, '-', splitLines(before)) // "@@ -1,N +0,0 @@"
	default:
		writeHunks(&b, splitLines(before), splitLines(after))
	}

	out := b.String()
	truncated := false
	if len(out) > maxDiffBytes {
		out = out[:maxDiffBytes]
		truncated = true
	}
	if out == "" && !truncated {
		return DiffResult{} // no-op edit: nothing changed
	}
	return DiffResult{UnifiedDiff: out, Kind: kind, Truncated: truncated}
}

// ReconcileDiff is ComputeUnifiedDiff under the plan's reconciliation name —
// the git-reconciliation path computes diffs with the identical engine.
func ReconcileDiff(before, after []byte, path string) DiffResult {
	return ComputeUnifiedDiff(before, after, path)
}

func kindFor(beforeExisted, afterExisted bool) string {
	switch {
	case !beforeExisted && afterExisted:
		return KindCreate
	case beforeExisted && !afterExisted:
		return KindDelete
	default:
		return KindModify
	}
}

// isBinary applies git's heuristic: a NUL byte in the first 8 KiB.
func isBinary(b []byte) bool {
	if len(b) > binaryProbeLen {
		b = b[:binaryProbeLen]
	}
	return bytes.IndexByte(b, 0) >= 0
}

// dline is one diffable line: its text (without the trailing newline) and
// whether the original content had a newline after it.
type dline struct {
	text   string
	hasEOL bool
}

// splitLines splits content into lines. A trailing "\n" does NOT produce an
// extra empty line — n lines = n "\n"-separated fragments, the convention
// unified diffs count with. The final fragment without a newline keeps
// hasEOL=false so the "\ No newline at end of file" marker can be emitted.
func splitLines(c []byte) []dline {
	if len(c) == 0 {
		return nil
	}
	out := make([]dline, 0, bytes.Count(c, []byte("\n"))+1)
	start := 0
	for i := 0; i < len(c); i++ {
		if c[i] == '\n' {
			out = append(out, dline{text: string(c[start:i]), hasEOL: true})
			start = i + 1
		}
	}
	if start < len(c) {
		out = append(out, dline{text: string(c[start:]), hasEOL: false})
	}
	return out
}

// op is one edit-script operation: ' ' keep, '-' remove, '+' add.
// For ' ' both indices are set; for '-' only aIdx; for '+' only bIdx.
type op struct {
	kind byte
	aIdx int
	bIdx int
}

// diffOps computes a minimal edit script via Myers' O((N+M)D) greedy
// algorithm (the same algorithm git diff uses). Both sides may be empty.
func diffOps(a, b []dline) []op {
	n, m := len(a), len(b)
	// Trim the common prefix/suffix first — the common case (a small edit in
	// a big file) then costs almost nothing.
	p := 0
	for p < n && p < m && a[p].text == b[p].text {
		p++
	}
	s := 0
	for s < n-p && s < m-p && a[n-1-s].text == b[m-1-s].text {
		s++
	}
	var ops []op
	for i := 0; i < p; i++ {
		ops = append(ops, op{kind: ' ', aIdx: i, bIdx: i})
	}
	ops = append(ops, opsMyers(a[p:n-s], b[p:m-s], p, s)...)
	for i := 0; i < s; i++ {
		ops = append(ops, op{kind: ' ', aIdx: n - s + i, bIdx: m - s + i})
	}
	return ops
}

// opsMyers runs Myers' greedy algorithm on the middle (non-trimmed) section
// and returns ops with absolute indices (prefix lines already consumed).
// A budget guard falls back to a naive whole-middle replace when the edit
// distance explodes (a near-total rewrite): Myers' O(D·(N+M)) work and
// per-round trace snapshots are unbounded otherwise, and for a near-total
// rewrite the naive "-"-all/"+"-all script IS what a minimal diff renders
// anyway. Correctness is unchanged — only minimality is given up, and the
// 256 KiB entry cap bounds the text either way.
func opsMyers(a, b []dline, prefix, suffix int) []op {
	n, m := len(a), len(b)
	if n == 0 && m == 0 {
		return nil
	}
	if n == 0 {
		ops := make([]op, 0, m)
		for i := range b {
			ops = append(ops, op{kind: '+', bIdx: prefix + i})
		}
		return ops
	}
	if m == 0 {
		ops := make([]op, 0, n)
		for i := range a {
			ops = append(ops, op{kind: '-', aIdx: prefix + i})
		}
		return ops
	}

	// Effort guard: work ≈ D·(N+M); cap trace snapshots ≈ 32 MiB.
	max := n + m
	offset := max
	dCap := (32 << 20) / (16*max + 8) // ints per round: (2max+1)·8B
	if dCap < 16 {
		dCap = 16
	}
	if max > 200_000 {
		return opsNaive(a, b, prefix)
	}
	v := make([]int, 2*max+1)
	var trace [][]int

find:
	for d := 0; d <= max; d++ {
		if d > dCap {
			return opsNaive(a, b, prefix)
		}
		snap := make([]int, len(v))
		copy(snap, v)
		trace = append(trace, snap)
		for k := -d; k <= d; k += 2 {
			var x int
			if k == -d || (k != d && v[offset+k-1] < v[offset+k+1]) {
				x = v[offset+k+1]
			} else {
				x = v[offset+k-1] + 1
			}
			y := x - k
			for x < n && y < m && a[x].text == b[y].text {
				x++
				y++
			}
			v[offset+k] = x
			if x >= n && y >= m {
				break find
			}
		}
	}

	// Backtrack the trace to reconstruct the (reversed) edit path.
	var rev []op
	x, y := n, m
	for d := len(trace) - 1; d >= 0; d-- {
		if d == 0 {
			// The origin IS the distance-0 endpoint — every op was already
			// consumed by rounds d>=1. Computing prevK against the zeroed
			// snapshot here would fabricate phantom ops (and negative
			// indices) for shared prefixes.
			continue
		}
		vv := trace[d]
		k := x - y
		var prevK int
		if k == -d || (k != d && vv[offset+k-1] < vv[offset+k+1]) {
			prevK = k + 1
		} else {
			prevK = k - 1
		}
		prevX := vv[offset+prevK]
		prevY := prevX - prevK
		for x > prevX && y > prevY { // diagonal (keep) steps
			x--
			y--
			rev = append(rev, op{kind: ' ', aIdx: x + prefix, bIdx: y + prefix})
		}
		if x == prevX && y == prevY {
			continue
		}
		if x > prevX { // horizontal step = removal from a
			x--
			rev = append(rev, op{kind: '-', aIdx: x + prefix})
		} else { // vertical step = addition from b
			y--
			rev = append(rev, op{kind: '+', bIdx: y + prefix})
		}
	}
	// The backtrack produced reversed ops; restore order.
	ops := make([]op, len(rev))
	for i := range rev {
		ops[i] = rev[len(rev)-1-i]
	}
	return ops
}

// opsNaive is the fallback edit script: remove the whole middle-A block,
// add the whole middle-B block. Linear time, always correct.
func opsNaive(a, b []dline, prefix int) []op {
	ops := make([]op, 0, len(a)+len(b))
	for i := range a {
		ops = append(ops, op{kind: '-', aIdx: prefix + i})
	}
	for i := range b {
		ops = append(ops, op{kind: '+', bIdx: prefix + i})
	}
	return ops
}

// writeHunks renders the ops into standard hunks (3 context lines, paired
// line numbers, "\ No newline at end of file" markers) following git's
// grouping rule: two changes land in the same hunk while the unchanged gap
// between them is ≤ 2*diffContext lines.
func writeHunks(w *bytes.Buffer, a, b []dline) {
	ops := diffOps(a, b)
	ctx := diffContext

	// Group the ops into hunks: walk the op list, opening a hunk at the
	// first change and extending it while another change arrives within
	// 2*ctx keeps of the previous one.
	type hunk struct{ from, to int } // [from, to) into ops
	var hunks []hunk
	i := 0
	for i < len(ops) {
		if ops[i].kind == ' ' {
			i++
			continue
		}
		// Open a hunk with up to ctx keeps before the first change.
		from := i - ctx
		if from < 0 {
			from = 0
		}
		to := i + 1
		// Extend while the next change is within 2*ctx of the last change.
		for to < len(ops) {
			if ops[to].kind != ' ' {
				to++
				continue
			}
			// Count keeps until the next change (or end).
			j := to
			for j < len(ops) && ops[j].kind == ' ' {
				j++
			}
			if j < len(ops) && j-to <= 2*ctx {
				to = j + 1 // absorb the keeps + the next change
				continue
			}
			break
		}
		// Close the hunk with up to ctx keeps after the last change.
		end := to + ctx
		if end > len(ops) {
			end = len(ops)
		}
		hunks = append(hunks, hunk{from: from, to: end})
		i = end
	}

	for _, h := range hunks {
		// Hunk header line numbers: the first A-consuming op fixes startA,
		// the first B-consuming op fixes startB. A hunk of only additions
		// (only deletions) starts at the A (B) cursor position with count 0
		// — git's convention, matching "@@ -0,0 +1,N @@" for creates.
		countA, countB := 0, 0
		var startA, startB int
		first := true
		for _, o := range ops[h.from:h.to] {
			switch o.kind {
			case ' ':
				countA++
				countB++
				if first {
					startA, startB = o.aIdx+1, o.bIdx+1
					first = false
				}
			case '-':
				countA++
				if first {
					startA, startB = o.aIdx+1, o.bIdx+1
					first = false
				}
			case '+':
				countB++
				if first {
					startA, startB = o.aIdx+1, o.bIdx+1
					first = false
				}
			}
		}
		// Git's convention: a count of 1 is printed bare ("-1", "+1"),
		// larger counts carry ",N" ("-1,3"). Verified against `git diff`.
		header := fmt.Sprintf("@@ -%s +%s @@\n",
			hunkRange(startA, countA), hunkRange(startB, countB))
		w.WriteString(header)
		for _, o := range ops[h.from:h.to] {
			switch o.kind {
			case ' ':
				w.WriteByte(' ')
				writeDline(w, a[o.aIdx])
			case '-':
				w.WriteByte('-')
				writeDline(w, a[o.aIdx])
			case '+':
				w.WriteByte('+')
				writeDline(w, b[o.bIdx])
			}
		}
	}
}

// writeWholeHunk renders the whole-file hunk for pure creates ('+' side,
// "@@ -0,0 +1,N @@") and pure deletes ('-' side, "@@ -1,N +0,0 @@").
func writeWholeHunk(w *bytes.Buffer, sign byte, lines []dline) {
	if sign == '+' {
		fmt.Fprintf(w, "@@ -0,0 +1,%d @@\n", len(lines))
	} else {
		fmt.Fprintf(w, "@@ -1,%d +0,0 @@\n", len(lines))
	}
	for _, l := range lines {
		w.WriteByte(sign)
		writeDline(w, l)
		if w.Len() > maxDiffBytes+1024 { // engine hard-stop; entry cap trims below
			break
		}
	}
}

// writeDline writes the line text plus newline, emitting the standard
// "\ No newline at end of file" marker for a final fragment without one.
func writeDline(w *bytes.Buffer, l dline) {
	w.WriteString(l.text)
	if l.hasEOL {
		w.WriteByte('\n')
		return
	}
	w.WriteString("\n\\ No newline at end of file\n")
}

// hunkRange renders one side of a hunk header: bare start when count==1
// (git's convention), "start,count" otherwise.
func hunkRange(start, count int) string {
	if count == 1 {
		return fmt.Sprintf("%d", start)
	}
	return fmt.Sprintf("%d,%d", start, count)
}
